### PR TITLE
Generate libSyntax API

### PR DESCRIPTION
**Pull request by harlanhaskins**
This patch replaces the hand-rolled implementation of libSyntax’s APIs with a set of Python definitions that drive a set of gyb templates.

The Node definitions are incomplete, and they will be filled out incrementally.

### CODE REVIEW

**Review by bitjammer — APPROVED** [review]

**Review by nkcsgexi — APPROVED** [review]

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by moiseev** [comment]
I have a question. Why mix JSON and gyb? gyb is already Python, and all the structures can be described in Python as well as in JSON. I think...

**Comment by harlanhaskins** [comment]
They can, but it feels a little weird describing the entire AST structure directly in python...it feels more like the job of a declarative language.

**Comment by moiseev** [comment]
Think about the benefits. You can compile your grammar definitions and even write tests for it! Just thinking out loud…

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test and merge

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test and merge

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

File: cmake/modules/AddSwift.cmake
Comment (gottesmm) [inline cmake/modules/AddSwift.cmake]: Why is this needed? Why doesn't the normal depends just work?

File: cmake/modules/AddSwift.cmake
Comment (gottesmm) [inline cmake/modules/AddSwift.cmake]: Typo. sorce -> source.

**Review by gottesmm — CHANGES_REQUESTED** [review]
I took a quick look at the cmake changes that you added. I have a question.

File: cmake/modules/AddSwift.cmake
Comment (harlanhaskins) [inline cmake/modules/AddSwift.cmake]: Thanks!

File: cmake/modules/AddSwift.cmake
Comment (harlanhaskins) [inline cmake/modules/AddSwift.cmake]: The normal `DEPENDS` won't regenerate gyb files if they're changed. This is actually no longer necessary for this PR, so I can remove it.

**Comment by gottesmm** [comment]
Also, generally we have not used gyb in the compiler itself. We have used it in the standard library. From what I remember adding gyb dependencies to the compiler is controversial. Can you bring this up for discussion on swift-dev? If we are going to add such a large dependency, we should make sure that no one has objections about it.

My main feeling with gyb is if it is possible to not gyb we should.

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Review by gottesmm — APPROVED** [review]

**Comment by gottesmm** [comment]
Harlan and I talked about this offline.

**Review by DougGregor — APPROVED** [review]
LGTM

**Comment by harlanhaskins** [comment]
@swift-ci please test

**Comment by harlanhaskins** [comment]
Rebased on master...

@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test and merge

**Comment by CodaFi** [comment]
@swift-ci please smoke test and merge

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
:boat:

File: include/swift/Syntax/CMakeLists.txt
Comment (dabrahams) [inline include/swift/Syntax/CMakeLists.txt]: I can't understand why you'd disable line directives like this; the result would be that any compilation errors point into the generated file instead of the original one, with the typical result that maintainers edit the wrong file.

Comment (harlanhaskins) [reply include/swift/Syntax/CMakeLists.txt]: Ahhh yes sorry about that. I turned line directives off while initially writing the gyb templates because it made it harder to see what the template instantiation was doing. That said, they should have been re-enabled before merging. I’ll re-enable them in a patch.

Comment (harlanhaskins) [reply include/swift/Syntax/CMakeLists.txt]: Now that I'm looking, these are for the C++ bindings. `gyb`'s line directive code emits Swift-style line directives -- clang will just ignore them. The patch for this would need to teach `gyb` how and when to emit C-style line directives, which wouldn't be too difficult.

Comment (dabrahams) [reply include/swift/Syntax/CMakeLists.txt]: Actually, those aren't swift-style line directives, which are broken for this purpose because they are only valid in certain syntactic positions.  These are just comments that get post-processed by the `line-directive` tool.  For error messages, it doesn't matter that clang would ignore them because the output of clang would get passed through the `line-directive` tool just like the output of Swift does.  However, generating real `#line` directives should also allow you to debug in the `.gyb` source rather than in the generated `.cpp`

**Pull request merged**
